source/tutorials/contributing-to-nixpkgs.md

```diff
@@ -0,0 +1,204 @@
+in
+{
+  icat = pkgs.callPackage ./icat.nix { };
+}
+```
+
+Copy both of these into `icat.nix` and `default.nix` files on your own machine, and then verify that you can build the derivation:
+
+```console
+$ nix-build -A icat
+...
+/nix/store/g6w508vxwr3df25dnl4k3xvcr4pqxprj-icat
+```
+
+The (omitted) output indicates that the package successfully built, and the very last line reports where Nix put the build result.
+
+## Phases and Hooks
+At the end of the previous tutorial, you learned that Nix derivations can be decomposed into a sequence of [phases](https://nixos.org/manual/nixpkgs/stable/#sec-stdenv-phases), each of which controls some aspect of the build process.
+
+You also learned about the existence of [*hooks*](https://nixos.org/guides/nix-pills/basic-dependencies-and-hooks), shell functions which execute in some of the derivation phases to control the build *environment*.
+
+Finally, you used this new knowledge to update the `installPhase` of the `icat` derivation with the pre- and post-install hooks:
+
+```diff
+# icat.nix
+...
+  installPhase = ''
++   runHook preInstall
+    mkdir -p $out/bin
+    cp icat $out/bin
++   runHook postInstall
+'';
+...
+```
+
+A number of other hooks are available, and when packaging for `nixpkgs` you should remember to include calls to these hooks in the phases you define, to make it easy for other contributors to [override](https://nixos.org/manual/nixpkgs/stable/#chap-overrides) parts of your derivations in the future.
+
+## Package Metadata
+The Nixpkgs [contribution guidelines](https://nixos.org/manual/nixpkgs/stable/#part-contributing) require all packages in Nixpkgs to have a `meta` attribute in their derivation, which contains information like a description of the package, the homepage of the project it belongs to, the software license, which platforms the package can be built for, and a list of Nixpkgs maintainers for the package.
+
+Usually, the required metadata can be gotten from the upstream repository.
+
+For `icat`, add the following `meta` attribute to the attrset passed to `stdenv.mkDerivation` in `icat.nix`:
+
+```nix
+# icat.nix
+...
+  meta = with lib; {
+    description = "icat (Image cat) outputs images in 256-color capable terminals.";
+    homepage = "https://github.com/atextor/icat";
+    license = licenses.bsdOriginal;
+    platforms = platforms.unix;
+  };
+...
+```
+
+With no further technical changes necessary, this derivation is now complete.
+
+## Adding to Nixpkgs
+You can now clone a local copy of `nixpkgs`, and move your derivation into an appropriate location within that source tree.
+
+```console
+$ git clone https://github.com/nixos/nixpkgs
+```
+
+`icat` is a console tool which also displays graphics, so there are a few places within Nixpkgs that you could reasonably move it.
+
+There are currently two other similar tools already in `nixpkgs`, both of which are stored in `pkgs/applications/graphics`, so this is a good candidate for a home for `icat` too.
```

Comment:
You should use [`pkgs/by-name`](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#readme) for new packages now! No need to edit `all-packages.nix` either that way :)